src/ucp/core/ucp_worker.c

```diff
@@ -1090,6 +1090,69 @@ ucp_worker_get_sys_device_distance(ucp_context_h context,
     return UCS_ERR_NO_RESOURCE;
 }
 
+static ucs_status_t
+ucp_worker_fill_uct_features(uct_md_h md, ucp_context_h context,
+                             uct_iface_params_t *iface_params)
+{
+    uint8_t mem_type_index = 0, has_gpu_memory = 0;
+    uct_md_attr_t md_attr;
+    ucs_status_t status;
+
+    status = uct_md_query(md, &md_attr);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    iface_params->field_mask |= UCT_IFACE_PARAM_FIELD_FEATURES;
+    iface_params->features    = 0;
+
+    ucs_for_each_bit(mem_type_index, md_attr.cap.reg_mem_types) {
+        if (UCP_MEM_IS_GPU(mem_type_index)) {
+            has_gpu_memory = 1;
+            break;
+        }
+    }
+
+    if (context->config.features & UCP_FEATURE_TAG) {
+        iface_params->features |= UCT_IFACE_FEATURE_TAG;
+    }
+
+    if (context->config.features & (UCP_FEATURE_AM | UCP_FEATURE_TAG | UCP_FEATURE_STREAM | UCP_FEATURE_RMA | UCP_FEATURE_AMO32 | UCP_FEATURE_AMO64)) {
+        iface_params->features |= UCT_IFACE_FEATURE_AM;
+    }
+
+    if (context->config.features & UCP_FEATURE_RMA) {
+        iface_params->features |= UCT_IFACE_FEATURE_PUT | UCT_IFACE_FEATURE_GET;
+    /*if (context->config.ext.proto_enable && context->num_mem_type_detect_mds > 0) {
+            iface_params->features |= UCT_IFACE_FEATURE_AM;
+        }*/
+    }
+
+    if (context->config.features & UCP_FEATURE_AMO32) {
+        iface_params->features |= UCT_IFACE_FEATURE_AMO32;
+    }
+
+    if (context->config.features & UCP_FEATURE_AMO64) {
+        iface_params->features |= UCT_IFACE_FEATURE_AMO64;
+    }
+
+    if ((context->config.ext.rndv_mode == UCP_RNDV_MODE_AUTO) ||
+    (context->config.ext.rndv_mode == UCP_RNDV_MODE_GET_ZCOPY) ||
+    (context->config.ext.rndv_mode == UCP_RNDV_MODE_GET_PIPELINE) ||
+    (context->config.ext.rndv_mode == UCP_RNDV_MODE_RKEY_PTR)) {
+        iface_params->features |= UCT_IFACE_FEATURE_GET;
+    }
+
+    if ((context->config.ext.rndv_mode == UCP_RNDV_MODE_PUT_ZCOPY) ||
+         (context->config.ext.rndv_mode == UCP_RNDV_MODE_PUT_PIPELINE) ||
```

Comment:
pipeline protocols need both get and put for memtype copy
so if have gpu memory both put and get have to be enabled